### ISSUE TITLE
[REVIEW]: imea: A Python package for extracting 2D and 3D shape measurements from images

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3091]
**Submitting author:** @nilskroell (<a href="http://orcid.org/0000-0002-6016-5144">Nils Kroell</a>)
**Repository:** <a href="https://git.rwth-aachen.de/ants/sensorlab/imea" target ="_blank">https://git.rwth-aachen.de/ants/sensorlab/imea</a>
**Version:** v0.3.1
**Editor:** @mbobra
**Reviewer:** @Atif-Anwer, @tuelwer
**Archive:** <a href="https://doi.org/10.5281/zenodo.4659161" target="_blank">10.5281/zenodo.4659161</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/b34442455601ccd0b474b814c417a531/status.svg)](https://joss.theoj.org/papers/b34442455601ccd0b474b814c417a531)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/b34442455601ccd0b474b814c417a531"><img src="https://joss.theoj.org/papers/b34442455601ccd0b474b814c417a531/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/b34442455601ccd0b474b814c417a531/status.svg)](https://joss.theoj.org/papers/b34442455601ccd0b474b814c417a531)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://git.rwth-aachen.de/ants/sensorlab/imea" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@Atif-Anwer & @tuelwer, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @mbobra know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://git.rwth-aachen.de/ants/sensorlab/imea">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@nilskroell) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://git.rwth-aachen.de/ants/sensorlab/imea">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@nilskroell) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @Atif-Anwer, @tuelwer it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.17 s (69.6 files/s, 23270.1 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                           7            557            752            890
Markdown                         2             66              0            197
TeX                              1             34              0            168
Jupyter Notebook                 2              0           1241            107
-------------------------------------------------------------------------------
SUM:                            12            657           1993           1362
-------------------------------------------------------------------------------


Statistical information for the repository '022b2085d0dde066d68f7c84' was
gathered on 2021/03/05.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Nils Kroell                     49          3725           3565           76.36
nkroell                          3          2148            109           23.64

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Nils Kroell                2199           59.0          0.1               12.73
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1016/j.mineng.2011.10.001 is OK
- 10.1016/j.wasman.2020.11.003 is OK
- 10.1016/j.ces.2013.02.043 is OK
- 10.1063/5.0002233 is OK
- 10.1109/IEMBS.2005.1617159 is OK
- 10.1016/s0165-0270(96)00080-5 is OK
- 10.1016/j.patrec.2017.08.022 is OK
- 10.1016/j.wasman.2020.09.033 is OK
- 10.7717/peerj.453 is OK
-  10.25080/Majora-92bf1922-00a is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03091/joss.03091/10.21105.joss.03091.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03091/joss.03091/10.21105.joss.03091.pdf) :page_facing_up: :point_left:

**Comment by mbobra** [comment]
👋  @Atif-Anwer @tuelwer Thank you so much for agreeing to review! You can find the article in the comment box above ⬆️ , and the checklist and software repository linked in the first comment box on this issue. We ask that you complete your review within six weeks, but let me know if you need more time. 

JOSS is an open review process and we encourage communication between the reviewers, the submitting author, and the editor. So please feel free to ask questions at any time, I'm always around.

I think you're good to go -- please let me know if you need anything else.

**Comment by tuelwer** [comment]
@mbobra @nilskroell This is my review for the submission:

The author proposes a Python library called imea which provides different functions to measure shapes in 2D or 3D images. Some of these functions are implemented from scratch, other wrap existing functions (from scikit-image and open-cv). After calculating the volume/area/surface measurements are returned as a pandas dataframe.

The toolbox provides a plethora of methods that calculate 2D and 3D shape measurements and is very accessible. I think the toolbox is of high value for researchers that have to calculate shape measurements from images on a regular basis.

I was able to install the library on MacOS 11.1 as well as on Debian 4.19 without problems using pip. 

 The repository provides two Jupyter notebooks containing an example application of the proposed functions (one notebook for 2D images and one for 3D image) and I was able to reproduce the stated results without problems. The example notebooks are very well-written and nicely demonstrate the easy usage of the toolbox.

Furthermore, I tested the package on an artificially generated image. The results were reasonable.

However, there are some issues that need to be addressed before I can recommend to accept this submission:
- **Documentation:** The source code contains doc-strings and the README.md gives an overview over the available functions and describes the structure of the toolbox. However, there is no more detailed documentation of the arguments of each function. What arguments are expected? What are the default values of these arguments? The only way to find this out is by scroll through the very long source code file and searching for the corresponding doc-string. This is very inconvenient.
- **Code quality:** The code is hard to read and does not follow the PEP 8 style guide. Function and variable names are unnecessary long and all code is contained in two Python files. The structure of the package could be highly improved by splitting these two very large Python files into multiple submodules (e.g., image preprocessing functions, calculating the actual measures, exporting to pandas dataframe.)
- **Unit tests:** The repository does not provide automated unit tests and there are no manual steps described to verify the functionality of the toolbox. I would recommend to write tests for each function and to use pytest to run the tests.
- **Installation from source:** It is not clear how to install the package after cloning the repository. Some instructions on that in the README.md would be useful.
- **Dependencies:** Although, the README.md states dependencies it is not clear which version of these packages needs to be installed. Also, it is common (for Python packages) to place a requirements.txt in the main folder that contains this information. This allows installation of the dependencies easily via a `pip install -r requirements.txt`. (See for example [scikit-image](https://github.com/scikit-image/scikit-image/blob/main/requirements/default.txt).)
- **References:** Some reference are missing in the README.md (e.g., (Stieß 2009), (Li et al. 2020)).
- **Terminology:** The term 3D image might be a little bit confusing: what exactly are a 3D height images? How do they differ from "regular" 2D images? I think a brief explanation (similar to the last sentence of (1) in `demo_3d_shape_measurements.ipynb`) would improve the README-file and the paper.
- **Missing warning:** The `extract_shape_measurements_2d` function returns an empty dataframe when no shapes have been found (for example when calling the function on `np.zeros((100, 100)).astype('bool')`). A warning would be useful in that case.

**Minor remarks:**
- README.md: "2d" vs. "2D"
- README.md: the ASCII art showing the coordinate system is broken. I suppose the "o" should mark the origin, is this correct?
- README.md (3D measurements): the code example should be part of a sentence (e.g., "The following function call can be used to...")
- README.md (code example): the function `extract_all_shape_measurements_2d`  was renamed to `extract_shape_measurements_2d`.

**Comment by nilskroell** [comment]
👋 @tuelwer,

thanks a lot for the detailed and very well written review. I really appreciate your invested time and effort for the review. These are all important points which will help to improve imea further. 

To adress all your points properly, I've created 10 new [issues ](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues) in the imea repository  and will work on them in the next days. I'll update you and give a detailed reply once these issues are fixed.

Best regards,
Nils

**Comment by mbobra** [comment]
@tuelwer Thank you so much for your detailed review! I really appreciate it.

**Comment by whedon** [comment]
:wave: @Atif-Anwer, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @tuelwer, please update us on how your review is going (this is an automated reminder).

**Comment by mbobra** [comment]
👋 @Atif-Anwer Please let me know where you're at so we can proceed with this review. (It is totally fine if you need more time). Also please let me know if you need any help -- I'm happy to help!

**Comment by nilskroell** [comment]
👋 @tuelwer,

I am more than happy to announce, that after several days of refactoring and coding I have finally addressed all your issues and released an substancial improvement to `imea` (now `v0.3.0`).

More specificially, the following points have been adressed:
- **Documentation** ([issue 48](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/48)): A HTML documentation including an API reference was added which is available at: [imea.readthedocs.io](https://imea.readthedocs.io/) (also linked in the [README](https://git.rwth-aachen.de/ants/sensorlab/imea)-file in the repository). Furthermore all docstrings have been refactored.
- **Code quality** ([issue 49](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/49)): Every line of code was refactored and the package was splitted into several (sub)modules. The organisation of submodules for the shape measurements is orientated on DIN ISO 9276-6, which suggests to group shape measurements into macro-, meso- and microdescriptors (for the 3D case only macrodecriptors are proprosed in the literature in technical norms). Variable and function names have been shortened and refactored, code is now consistent with PEP 8. `autopep8`, `pylint` and `flake8` was used to validate and improve code quality.
- **Unit tests** ([issue 50](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/50)): As suggested, automated unit tests for all relevant functions (e.g. individual shape measurements, functions for providing shape measurement as `pandas` dataframe) were added into a new tests folder. `pytest` was used to run and evaluate all tests, with the submitted version (`v0.3.0`) all tests are passing.
- **Installation from source** ([issue 51](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/51)): An instruction how to install `imea` from source has been added to the [documentation](https://imea.readthedocs.io/) and [README](https://git.rwth-aachen.de/ants/sensorlab/imea)-file.
- **Dependencies** ([issue 52](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/52)): A `requirements.txt` has been added to the repository and the versions are specified in the [documentation](https://imea.readthedocs.io/) and [README](https://git.rwth-aachen.de/ants/sensorlab/imea)-file.
- **References** ([issue 53](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/53)): The missing references has been added (see [README](https://git.rwth-aachen.de/ants/sensorlab/imea)).
- **Terminology** ([issue 54](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/54) & [issue 55](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/55)): The terminology has been specified in paper and [README](https://git.rwth-aachen.de/ants/sensorlab/imea).
- **Missing warning** ([issue 56](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/56)): A warning has been added when no objects are found for the 2D and 3D case. Furthermore, warnings have been implemeted for missing shapes in the functions calculating the individual shape measurements (also see tests).
- **Minor remarks** ([issue 57](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/57)): Minor remarks regarding the [README](https://git.rwth-aachen.de/ants/sensorlab/imea)-file have been fixed.

If you have further questions, I am happy to answer and would like to thanks you once again for the detailed review.

Best regards,
Nils

**Comment by nilskroell** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

```

**Comment by nilskroell** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03091/joss.03091/10.21105.joss.03091.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03091/joss.03091/10.21105.joss.03091.pdf) :page_facing_up: :point_left:

**Comment by nilskroell** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1016/j.mineng.2011.10.001 is OK
- 10.1016/j.wasman.2020.11.003 is OK
- 10.1016/j.ces.2013.02.043 is OK
- 10.1063/5.0002233 is OK
- 10.1109/IEMBS.2005.1617159 is OK
- 10.1016/s0165-0270(96)00080-5 is OK
- 10.1016/j.patrec.2017.08.022 is OK
- 10.1016/j.wasman.2020.09.033 is OK
- 10.7717/peerj.453 is OK
-  10.25080/Majora-92bf1922-00a is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by nilskroell** [comment]
@whedon check repository

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.09 s (501.6 files/s, 66022.8 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          25            701           1326           1588
Markdown                        12            105              0            351
TeX                              1             34              0            168
Jupyter Notebook                 2              0           1230            107
YAML                             2              6              5             31
JSON                             1              0              0              8
-------------------------------------------------------------------------------
SUM:                            43            846           2561           2253
-------------------------------------------------------------------------------


Statistical information for the repository '5972d6fabb4192ff30e19931' was
gathered on 2021/03/26.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Nils Kroell                    183         11030           9454           90.08
nkroell                          3          2148            109            9.92

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Nils Kroell                3615           32.8          0.1                8.27
```

**Comment by Atif-Anwer** [comment]
Absolutely sorry for being awol for so long. Just got really bogged down in IRL deadline and had no time to do anything else. Ill update my review by tomorrow for sure since im looking at the repo right now. Apologies again for the delay. 
Cheers!

**Comment by Atif-Anwer** [comment]
This is my review for the paper:

The submission is a Python library titled "imea" that can extract different measurements of 2D as well as 3D shapes. Some functions are new implementations whereas some functions are wrappers around traditional functions from various libraries. The package combines these functions in an easy to use library allowing the user to calculate a huge range of statistics from a single package.
Overall, the package is well documented and available at imea.readthedocs.io, with a clear and simple ways to install and use the package. 
The tests were done on Ubuntu 20.04.2 LTS installation without any issues. The provided jupyter notebooks and examples are clear and well documented for the user to follow aswell. 
Testing was also done on custom images other than the ones provided in the demo aswell with good results whenever the image was successfully processed. 

Overall, (and especially after the revisions by the author as documented here (https://github.com/openjournals/joss-reviews/issues/3091#issuecomment-808522623), I found no major issues with toolbox, and any issues there were have been addressed already based on the suggestions of @tuelwer .  There are a few minor suggestions to the provided example and the paper however, that if addressed will help improve user experience and the paper. 

1. The provided 3D example is based on data provided in csv files but the package mentions processing 3D images. It would be better if the author can add a demo depth image (png from a Kinect/Realsense etc) and its implementation as an example so that the users can follow that easily also.
 
2. The submitted paper only contains a 'statement of need' section. However, it might be nice to add an example image with some extracted statistics in the Summary section. Also, a table or appendix (if allowed in the JOSS format) with the included statistics that are extracted) will also be nice, making it akin to a conventional publication and help others cite the publication when it is used in any work. 

Other than that, its a well rounded package, and I really have no issues with the implementations, code or performance etc.Great job!

**Comment by tuelwer** [comment]
I appreciate the author's effort to improve and restructure the toolbox! A majority of the issues that are stated above are addressed in an appropriate way and I checked some of the missing check-boxes. However, there are a couple of minor issues that should be fixed before publication:
- [ ] There are no instructions on how to run the new tests. I suggest that the author writes a short section in the docs and/or in the README.md on how to run the tests. (In essence, the command `python3 -m pytest tests` should be mentioned.) Also pytest could be mentioned as a requirement (in a similar way like matplotlib is mentioned.)
- [ ] There are warnings when running the tests. The FutureWarnings (indicating deprecation) should be fixed and the UserWarnings could be caught while running the tests using [this context manager](https://docs.python.org/3/library/warnings.html#testing-warnings) when they are expected.
- [ ] I like the new sub-modules that have been defined for the shape measurements! I suggest that the author include a brief explanation for each of the groups (macro-, meso- and microdescriptors) in the documentation. This would make it easier for the user to find the right measure.

Besides these points, I see no more issues with this submission!

**Comment by nilskroell** [comment]
👋 @Atif-Anwer,

thanks a lot for your invested time and the nice review. I have adressed your issues as follows:

> 1. The provided 3D example is based on data provided in csv files but the package mentions processing 3D images. It would be better if the author can add a demo depth image (png from a Kinect/Realsense etc) and its implementation as an example so that the users can follow that easily also.

**Issue [#71](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/71)**: I converted the .csb-images to 16bit png files and adjusted the spatial resolution in z-direction, accordingly. Furthermore, the demo notebook has been updates to read the 3D images from the png files. 

> 2. The submitted paper only contains a 'statement of need' section. However, it might be nice to add an example image with some extracted statistics in the Summary section. Also, a table or appendix (if allowed in the JOSS format) with the included statistics that are extracted) will also be nice, making it akin to a conventional publication and help others cite the publication when it is used in any work.

**Issue [#72](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/72)**: I've added a figure to the paper that shows an exemplary selection of some 2D shape measurements for better understanding. As the shape measurements are already listed in the [documentation](https://imea.readthedocs.io/) and [README](https://git.rwth-aachen.de/ants/sensorlab/imea/-/blob/master/README.md) file, I decided against adding an appendix with available shape measurements  in the paper. There are two reasons for that:
(a) It is likely that in the future more shape measurements might be added to `imea`, in this case the paper will be outdated and might confuse the reader/user.
(b) As the [JOSS paper guidelines](https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain) suggest, the aim of this paper is to give "[a] summary describing the high-level functionality and purpose of the software for a diverse, non-specialist audience.". Regarding this I think that providing a list of all shape measurements might be too detailed for the purpose of the paper, and the documentation is a better place. However, I added some more explaination to the documentation and README file, so that it will be even better understandable for users.

Thanks again for your invested time and review, I appreciate it a lot.

If there are further questions, I am happy to answer. If not, I would be happy, if you can check, if all your comments have been adressed properly.

Best regards,
Nils

**Comment by nilskroell** [comment]
👋 @tuelwer ,

thanks for your answer and the follow up. I have adressed the remaining issues as follows:

> There are no instructions on how to run the new tests. I suggest that the author writes a short section in the docs and/or in the README.md on how to run the tests. (In essence, the command python3 -m pytest tests should be mentioned.) Also pytest could be mentioned as a requirement (in a similar way like matplotlib is mentioned.)

**Issue [#76](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/76):** I have added a short section how to run the test to the [README](https://git.rwth-aachen.de/ants/sensorlab/imea/-/blob/master/README.md) and [documentation](imea.readthedocs.io) and `pytest` was added as an requirement for testing, as suggested. Furthermore, I created a `requirements-dev.txt` that bundles all requirements for testing and developing (as e.g. also done by [pandas](https://github.com/pandas-dev/pandas)).

> There are warnings when running the tests. The FutureWarnings (indicating deprecation) should be fixed and the UserWarnings could be caught while running the tests using this context manager when they are expected.

**Issue [#77](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/77):** The future warnings have been fixed. (I used `scikit-image v0.16` and therefore did not get any warnings. However, after updating to 0.19, I saw the warnings and adapted the code accordingly.)

**Issue [#79](https://git.rwth-aachen.de/ants/sensorlab/imea/-/issues/79):** The user warning have been catched and checked using the context manager you suggested. This was a great tip, thanks! Furthermore, I have also adjusted some warnings, s.t. they are more consistent throughout the package.

> I like the new sub-modules that have been defined for the shape measurements! I suggest that the author include a brief explanation for each of the groups (macro-, meso- and microdescriptors) in the documentation. This would make it easier for the user to find the right measure.

Thanks for the feedback, I am also very happy with it 😄. As suggested, I have added an explanation as well as an additional illustration to the [README](https://git.rwth-aachen.de/ants/sensorlab/imea/-/blob/master/README.md)-file and [documentation](imea.readthedocs.io). Furthermore, I have added some examples and explaination to the paper as suggested by @Atif-Anwer .

I would like to thank you again for your invested time and effort, I truely learned a lot by implementing your suggestions.

I would be happy, if you can check if all your comments have been implemented successfully. If there are further questions, I am happy to answer.

Best regards,
Nils

**Comment by nilskroell** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03091/joss.03091/10.21105.joss.03091.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03091/joss.03091/10.21105.joss.03091.pdf) :page_facing_up: :point_left:

**Comment by mbobra** [comment]
@nilskroell Thank you for such a quick follow-up!

@Atif-Anwer @tuelwer Did the submitting author address all of your concerns? If so I will move forward with accepting the paper. Please let me know!

**Comment by tuelwer** [comment]
@nilskroell, thanks for you fast response to my feedback! I can confirm that the tests now run without any warnings on my machine. Also, I think that the new explanations make things more clear. Good job! 👍

@mbobra, I see no more issues with this submission and recommend to accept it 👍

**Comment by nilskroell** [comment]
@tuelwer Thanks, for the quick follow up. Thats great to hear! Thanks once again for the great review(s) and responses, I really appreciate it. 😀 

@mbobra : Great to hear! Can we update the version number listed above to `v0.3.1` (the current version of `imea` after adressing the reviewers concerns)?

**Comment by mbobra** [comment]
@nilskroell Yes, perfect! If @Atif-Anwer gives us the thumbs up, then please release a new version of `imea` and deposit the code in Zenodo. [Here are some instructions](https://guides.github.com/activities/citable-code/) about how to deposit code from Github to Zenodo, but I am not sure if it works with Gitlab. Please make sure the author list on the Zenodo deposit matches the author list on the paper.

**Comment by Atif-Anwer** [comment]
@nilskroell Great work! I can also confirm that all suggestions have been incorporated the amendments run without any issue, including updated paper draft. I do think it is much more clear and citable now :) 

@mbobra All remarks and comments have been successfully addressed therefore I would like to recommend the submission for acceptance and publishing 👍🏼

**Comment by nilskroell** [comment]
@Atif-Anwer Thats great to hear! Thanks a lot for your great review and the fast reply -- it was a pleasure! 👍

**Comment by nilskroell** [comment]
@mbobra The release version for JOSS (v0.3.1) has been added as a tag on the `imea` repository: [`v0.3.1`](https://git.rwth-aachen.de/ants/sensorlab/imea/-/releases/v0.3.1).
Furthermore I archived v0.3.1 on Zenodo, the DOI is: [10.5281/zenodo.4659161](https://www.doi.org/10.5281/zenodo.4659161) The other list in Zenodo is the same as in this review (just myself with included ORCID tag)
*(Just as a sidenote: The integration of GitLab in the same convient way as GitHub is currently not supported by Zenodo (see this [issue](https://github.com/zenodo/zenodo/issues/1404) for details); however I uploaded the .zip-file of `v0.3.1` to Zenodo have a full archive)*

Is there anything else I can do or help with? 😄

**Comment by mbobra** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03091/joss.03091/10.21105.joss.03091.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03091/joss.03091/10.21105.joss.03091.pdf) :page_facing_up: :point_left:

**Comment by mbobra** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1016/j.mineng.2011.10.001 is OK
- 10.1016/j.wasman.2020.11.003 is OK
- 10.1016/j.ces.2013.02.043 is OK
- 10.1063/5.0002233 is OK
- 10.1109/IEMBS.2005.1617159 is OK
- 10.1016/s0165-0270(96)00080-5 is OK
- 10.1016/j.patrec.2017.08.022 is OK
- 10.1016/j.wasman.2020.09.033 is OK
- 10.7717/peerj.453 is OK
-  10.25080/Majora-92bf1922-00a is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by mbobra** [comment]
@whedon set 10.5281/zenodo.4659161 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4659161" target="_blank">10.5281/zenodo.4659161</a> is the archive.

**Comment by mbobra** [comment]
@whedon set v0.3.1 as version

**Comment by whedon** [comment]
OK. v0.3.1 is the version.

**Comment by mbobra** [comment]
@nilskroell Congratulations, your paper has been accepted to JOSS! 🎉

@tuelwer @Atif-Anwer Thank you so very much for volunteering to review this submission. I really appreciate your time and care ☀️

@openjournals/jose-eics This paper is ready for final processing.

**Comment by danielskatz** [comment]
Sorry to be a little harsh, but this paper has not yet been accepted.  The process continues...

@mbobra - your next step is to run `@whedon accept`, make sure everything looks good, then I (or the AEiC on duty) does another proof-read and continue the last steps of processing.

**Comment by mbobra** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by mbobra** [comment]
@openjournals/jose-eics This paper is ready for final processing.

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1016/j.mineng.2011.10.001 is OK
- 10.1016/j.wasman.2020.11.003 is OK
- 10.1016/j.ces.2013.02.043 is OK
- 10.1063/5.0002233 is OK
- 10.1109/IEMBS.2005.1617159 is OK
- 10.1016/s0165-0270(96)00080-5 is OK
- 10.1016/j.patrec.2017.08.022 is OK
- 10.1016/j.wasman.2020.09.033 is OK
- 10.7717/peerj.453 is OK
-  10.25080/Majora-92bf1922-00a is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2186

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2186, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
👋 @openjournals/dev - the xml file here seems to be missing almost all of the references.  Can you look into this?

**Comment by danielskatz** [comment]
@nilskroell - there are some small changes needed in the paper, but I am unable to fork it in GitLab, so I'm just attaching here a new version that has the changes.  Also, I can't attach an .md file here, so this is called .txt, but it's really md
[paper.txt](https://github.com/openjournals/joss-reviews/files/6260263/paper.txt)

**Comment by nilskroell** [comment]
@danielskatz Thanks a lot for editing and improving the paper, I commited your changes with the latest commit. 👍

**Comment by nilskroell** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03091/joss.03091/10.21105.joss.03091.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03091/joss.03091/10.21105.joss.03091.pdf) :page_facing_up: :point_left:

**Comment by nilskroell** [comment]
@danielskatz @mbobra I have proof read the paper again. Thanks a lot for your improvements and quick responses! From my side, everything is ready for publication.

**Comment by danielskatz** [comment]
Thanks - before we can finish up, we need @openjournals/dev to help with the lack of references in the XML file.

**Comment by arfon** [comment]
The issue here is that Whedon doesn't know how to handle BibTeX keys with periods in them (e.g., https://git.rwth-aachen.de/ants/sensorlab/imea/-/blob/master/paper/paper.bib#L1. `Andersson.2012`). 

I can process this paper manually if it's ready to go, or if @nilskroell is inclined to remove the periods in the paper and the BibTeX file then we can have Whedon accept automatically here.

**Comment by nilskroell** [comment]
@arfon Thanks a lot for the quick response. I fix the BibTeX keys an pushed everything to `master` branch. Now it should work @danielskatz .

**Comment by nilskroell** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03091/joss.03091/10.21105.joss.03091.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03091/joss.03091/10.21105.joss.03091.pdf) :page_facing_up: :point_left:

**Comment by nilskroell** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1016/j.mineng.2011.10.001 is OK
- 10.1016/j.wasman.2020.11.003 is OK
- 10.1016/j.ces.2013.02.043 is OK
- 10.1063/5.0002233 is OK
- 10.1109/IEMBS.2005.1617159 is OK
- 10.1016/s0165-0270(96)00080-5 is OK
- 10.1016/j.patrec.2017.08.022 is OK
- 10.1016/j.wasman.2020.09.033 is OK
- 10.7717/peerj.453 is OK
-  10.25080/Majora-92bf1922-00a is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by nilskroell** [comment]
@danielskatz @mbobra After checking everything again, I can confirm that from my side everything is go for publication.

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1016/j.mineng.2011.10.001 is OK
- 10.1016/j.wasman.2020.11.003 is OK
- 10.1016/j.ces.2013.02.043 is OK
- 10.1063/5.0002233 is OK
- 10.1109/IEMBS.2005.1617159 is OK
- 10.1016/s0165-0270(96)00080-5 is OK
- 10.1016/j.patrec.2017.08.022 is OK
- 10.1016/j.wasman.2020.09.033 is OK
- 10.7717/peerj.453 is OK
-  10.25080/Majora-92bf1922-00a is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2193

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2193, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1379413952015314947) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2194
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.03091](https://doi.org/10.21105/joss.03091)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by danielskatz** [comment]
Congratulations to @nilskroell (Nils Kroell)!!

And thanks to @Atif-Anwer and @tuelwer for reviewing, and @mbobra for editing!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03091/status.svg)](https://doi.org/10.21105/joss.03091)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03091">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03091/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03091/status.svg
   :target: https://doi.org/10.21105/joss.03091
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03091/status.svg)](https://doi.org/10.21105/joss.03091)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by nilskroell** [comment]
Thanks a lot to the reviewers @tuelwer and @Atif-Anwer for your valuable feedback, @mbobra and @danielskatz for the submission support and editing and @arfon for the technical support at the end.
It was a very valuable experience to publish a paper into JOSS and I would like to thank everybody who contributed to this wonderful experience. 😄
